### PR TITLE
[RISCV][NFC] Allow SchedVar to be a def inside our scheduler model files.

### PR DESCRIPTION
**Pull request by michaelmaitland**
All SchedModel files have a line that looks like:

```
def SomeModel : SchedMachineModel;
let SchedModel = SomeModel in {
  ...
}
```

TableGen requires that all records defined within the top level `let` must have a field `SchedModel` somewhere in their nested record hierarchy (i.e. the record has a field `SchedModel : SchedMachineModel` or recursively, one of its members has a field `SchedModel : SchedMachineModel`).

Classes such as `SchedPredicate` have added a field `SchedModel : SchedMachineModel`, even though the field is never used, just to supress **errors** (not warnings) caused from having the top level let in the model files. This decision was made to avoid having hundreds of the same `let` statement littered in every scheduler model file.

The reason we have never seen an error for `SchedVar` before is because `SchedVar` is never instantiated with a `def`. Instead, it is only created as a value that is consumed by `SchedWriteVariant`:

```
... : SchedWriteVariant<[SchedVar<...>, SchedVar<...>]>;
```

There is a problem with this style of instantiation. In particular, the problem arises as we try to take a class based approach to building scheduler models. I will describe the problem from the bottom up.

The `LMULWriteResMXVariant` multiclass takes in a `SchedPredicateBase Pred`. Today, the RISCVSchedSiFive7.td file defines `VLDSX0Pred` outside the scope of any class. That means that `VLDSX0Pred` exists before `LMULWriteResMXVariant` multiclass is instantiated. With this approach, there is no error since the predicate is instantated in entirety before the variant multiclass is instantiated. However, I have the intention to move the definition of both the predicate and the variant multiclass records inside a multiclass to factor out common parts between multiple scheduler models.

I plan to have something like:

```
multiclass SiFive7Base<SiFive7BaseConfig c> {
  def VLDSX0Pred : ...;
  // Need defvar since record is prefixed with NAME.
  defvar VLDSX0Pred = !cast<...>(NAME # VLDSX0Pred);
  defm SiFive7 : LMULWriteResMXVariant<VLDSX0Pred>;
}

defm "SiFive7Version1" : SiFive7Base<SiFive7BaseConfig<...>>;
defm "SiFive7Version2" : SiFive7Base<SiFive7BaseConfig<...>>;
```

In this scheme, VLDSX0Pred is defined within the same multiclass transaction that the `LMULWriteResMXVariant` is defined in. For some reason, TableGen does not allow `Values` to reference records that were created in the same parent record construction. If the `SchedVar` is not a `def`, then it will not be able to find the record `NAME # VLDSX0Pred`. Making it a def, allows TableGen to find `NAME # VLDSX0Pred` in scope.

The simplest example of this is:

```
class A {}
class B<A a> { A x = a;}
class C<B b> { B y = b;}
multiclass D {
  def MyA : A;
  defvar aa = !cast<A>(NAME # MyA);
  // This works
  def : B<aa>;
  // This does not work because constructing B by value cannot find `NAME # MyA`
  // error: Undefined reference to record: 'MyA'
  def : C<B<aa>>;
  // To fix it, define it like such:
  def MyB : B<aa>;
  defvar bb = !cast<B>(NAME # MyB);
  def : C<bb>;
}
defm "" : D;
```

In summary, in order to use a class based approach to creating scheduler resources to promote resusability, `SchedVar`s must be created using defs instead of being instantiated by value so that it can resolve records that were part of the instantiation of the parent record being created. In order to do this without refactoring the top level `let` statement that all scheduler model files use, we add an unused field `SchedModel : SchedMachineModel` to `SchedVar`, similiar to what has been done in `SchedPredicate`.

### CODE REVIEW

**Comment by topperc** [comment]
Why does `VLDSX0Pred` need to be inside the SiFive7Base class? There's nothing that changes based on which model uses it. It could just be a top level record.

**Comment by michaelmaitland** [comment]
> Why does `VLDSX0Pred` need to be inside the SiFive7Base class? There's nothing that changes based on which model uses it. It could just be a top level record in RISCVScheduleV.td. ARM, AArch64, and X86 have a SchedPredicates.td file. I don't think we have enough predicates yet to justify that, but my point is that sched predicates can be independent of the model.

`SchedPredicate` class has access to the SchedModel. For example, it would be okay to say:
```
def MyPred : SchedPredicate<[{SchedModel->getSubtargetInfo()->hasFeature(RISCV::TuneMyFeature)}]>;
```
In these cases, the predicate needs to be defined under the `let` statement.

**Comment by topperc** [comment]
> > Why does `VLDSX0Pred` need to be inside the SiFive7Base class? There's nothing that changes based on which model uses it. It could just be a top level record in RISCVScheduleV.td. ARM, AArch64, and X86 have a SchedPredicates.td file. I don't think we have enough predicates yet to justify that, but my point is that sched predicates can be independent of the model.
> 
> `SchedPredicate` class has access to the SchedModel. For example, it would be okay to say:
> 
> ```
> def MyPred : SchedPredicate<[{SchedModel->getSubtargetInfo()->hasFeature(RISCV::TuneMyFeature)}]>;
> ```
> 
> In these cases, the predicate needs to be defined under the `let` statement.

But we're talking about an MCSchedPredicate here not a SchedPredicate. Does an MCSchedPredicate have access to the model?

**Comment by topperc** [comment]
> > Why does `VLDSX0Pred` need to be inside the SiFive7Base class? There's nothing that changes based on which model uses it. It could just be a top level record in RISCVScheduleV.td. ARM, AArch64, and X86 have a SchedPredicates.td file. I don't think we have enough predicates yet to justify that, but my point is that sched predicates can be independent of the model.
> 
> `SchedPredicate` class has access to the SchedModel. For example, it would be okay to say:
> 
> ```
> def MyPred : SchedPredicate<[{SchedModel->getSubtargetInfo()->hasFeature(RISCV::TuneMyFeature)}]>;
> ```
> 
> In these cases, the predicate needs to be defined under the `let` statement.

Why? That SchedModel is coming from C++ code not from tablegen. It does not involve the let statement at all.

**Comment by michaelmaitland** [comment]
We can move the predicate out, that is unrelated to this patch. We still need to def the SchedVar, otherwise it wont be able to find the SchedReadWrite that was defined just a few lines above.

**Review by topperc — APPROVED** [review]
LGTM

**Review by wangpc-pp — APPROVED** [review]
LGTM.

> For some reason, TableGen does not allow Values to reference records that were created in the same parent record construction.

Yeah, this is a problem. I may make TableGen able to refer to other definitions in a multiclass:
```
multiclass Test {
  def a;
  def b: B<a>; // For `a`, we will search `NAME # a` first and then `a`.
}
```
(Hopefully this can be easy to design and implement)

**Comment by michaelmaitland** [comment]
> Yeah, this is a problem. I may make TableGen able to refer to other definitions in a multiclass:
> 
> ```
> multiclass Test {
>   def a;
>   def b: B<a>; // For `a`, we will search `NAME # a` first and then `a`.
> }
> ```
> 
> (Hopefully this can be easy to design and implement)

This would be very nice. It is quite ugly to have to use `!cast` everywhere for defs. Some careful though will need to be given to what happens when there is a record `a` and when there is a different record `NAME # a` -- we would need a way to refer to both. Please include me on the review if you do this work :)

**Comment by wangpc-pp** [comment]
> > Yeah, this is a problem. I may make TableGen able to refer to other definitions in a multiclass:
> > ```
> > multiclass Test {
> >   def a;
> >   def b: B<a>; // For `a`, we will search `NAME # a` first and then `a`.
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > (Hopefully this can be easy to design and implement)
> 
> This would be very nice. It is quite ugly to have to use `!cast` everywhere for defs. Some careful though will need to be given to what happens when there is a record `a` and when there is a different record `NAME # a` -- we would need a way to refer to both. Please include me on the review if you do this work :)

I have tried it, it may be hard to implement based on my rough investigation. In short, the TableGen parser resolves Record after resolving variables/arguements. I need more time to think about it.
